src/tpm/tpm_wrapper.rs

```diff
@@ -1,240 +1,204 @@
+use libc::c_void;
+use std::{ffi::CString, mem::MaybeUninit, ptr::null_mut, sync::Mutex};
 use tpm::Error as TpmError;
-
 use tss2::{
     Esys_ContextLoad, Esys_ECDH_ZGen, Esys_Finalize, Esys_FlushContext, Esys_Initialize,
     Esys_ReadPublic, Fapi_Finalize, Fapi_GetEsysBlob, Fapi_GetTcti, Fapi_Initialize, Fapi_Sign,
     Tss2_MU_TPMS_CONTEXT_Unmarshal, ESYS_CONTEXT, ESYS_TR, ESYS_TR_NONE, ESYS_TR_PASSWORD,
-    FAPI_CONTEXT, FAPI_ESYSBLOB_CONTEXTLOAD, TPM2B_ECC_POINT, TPM2B_NAME, TPM2B_PUBLIC,
-    TPMS_CONTEXT, TSS2_RC_SUCCESS, TSS2_TCTI_CONTEXT, UINT16,
+    FAPI_CONTEXT, FAPI_ESYSBLOB_CONTEXTLOAD, TPM2B_ECC_POINT, TPM2B_PUBLIC, TPMS_CONTEXT,
+    TSS2_RC_SUCCESS, TSS2_TCTI_CONTEXT, UINT16,
 };
 
 pub type Result<T = ()> = std::result::Result<T, error::Error>;
 
-static mut TPM_CTX: Option<Mutex<*mut FAPI_CONTEXT>> = None;
-
-fn tpm<'a>() -> &'a Mutex<*mut FAPI_CONTEXT> {
-    unsafe { TPM_CTX.as_ref().unwrap() }
+// A wrapper which wraps TSS2 APIs and handles converting the return
+// code from an integer to a `Result`.
+macro_rules! tss2_call{
+    ( $func:ident ( $( $arg:expr ),* $(,)? ) ) => {{
+        match $func($($arg),*) {
+            TSS2_RC_SUCCESS => Ok(()),
+            e => Err(TpmError::TPMError(stringify!($func), e)),
+        }
+    }};
 }
 
-fn with_tpm<F, R>(f: F) -> R
-where
-    F: FnOnce(*mut FAPI_CONTEXT) -> R,
-{
-    let tpm_ctx = tpm().lock().unwrap();
-    f(*tpm_ctx)
-}
+/// An RAII wrapper for FAPI_CONTEXT.
+struct FapiContext(*mut FAPI_CONTEXT);
 
-pub fn tpm_init() -> Result {
-    unsafe {
+impl FapiContext {
+    pub fn new() -> Result<Self> {
         let mut tpm_ctx: *mut FAPI_CONTEXT = null_mut();
-        let res = Fapi_Initialize(&mut tpm_ctx as *mut *mut FAPI_CONTEXT, null_mut());
-        if res != TSS2_RC_SUCCESS {
-            return Err(TpmError::tpm_error("Fapi_Initialize", res));
-        }
-
-        TPM_CTX = Some(Mutex::new(tpm_ctx));
+        unsafe {
+            tss2_call!(Fapi_Initialize(
+                &mut tpm_ctx as *mut *mut FAPI_CONTEXT,
+                null_mut(),
+            ))?;
+        };
+        Ok(Self(tpm_ctx))
     }
-
-    Ok(())
 }
 
-pub fn tpm_deinit() {
-    unsafe {
-        let mut tpm_ctx = *tpm().lock().unwrap();
-        Fapi_Finalize(&mut tpm_ctx);
-        TPM_CTX = None;
-    }
-}
+unsafe impl Send for FapiContext {}
 
-fn free_esys_resources(esys_ctx: &mut *mut ESYS_CONTEXT, esys_key_handle: ESYS_TR) -> Result {
-    unsafe {
-        if esys_key_handle != u32::MAX {
-            let res = Esys_FlushContext(*esys_ctx, esys_key_handle);
-            if res != TSS2_RC_SUCCESS {
-                return Err(TpmError::tpm_error("Esys_FlushContext", res));
-            }
+impl Drop for FapiContext {
+    fn drop(&mut self) {
+        unsafe {
+            Fapi_Finalize(&mut self.0);
         }
-
-        Esys_Finalize(&mut *esys_ctx);
     }
+}
 
-    Ok(())
+lazy_static! {
+    static ref TPM_CTX: Mutex<FapiContext> = Mutex::new(FapiContext::new().unwrap());
```

Comment:
TODO: I'd want to get rid of this `unwrap()`, but working with a global `Result<Mutex>`/`Mutex<Result>` was slowing progress.